### DISCUSSION TITLE
Modules inside of Modules?

### DISCUSSION THREAD

**Discussion by MattDRogers** [netbox-community/netbox#14124] [Q&A] [answered]
I have run into a modeling scenario with my routers that I am sure others have as well. I'd like to see how people are approaching this.
I'll give some examples of what I'm talking about:

**Juniper MX** router modular line cards (Modules) such as the MPC2E-3D-NG are designed to house up to two smaller line cards (also Modules) called MICs. 
**Cisco ASR 9K** router modular line cards (Modules) such as the A9K-MOD400-TR are designed to house up to two smaller line cards (also Modules) called MPAs.
**Nokia 7750 SR-Series** router modular line cards (Modules) such as the IOM3-XP are designed to house up to two smaller line cards (also Modules) called MDAs.

I seem to be able to assign multiple Modules to a single Module Bay Position--which is cool. But, I'd need a variable like we use for {module} in the "sub-slot" position.

I'm a visual guy, so here's some pictures of what I'm talking about:
![image](https://github.com/netbox-community/netbox/assets/149083057/5c37b64e-32a6-49dd-9748-71efea1db8c4)
![image](https://github.com/netbox-community/netbox/assets/149083057/ee7c9141-b415-4c89-9433-a89bd14e3285)
![image](https://github.com/netbox-community/netbox/assets/149083057/5c119132-d09a-4af7-9e90-7b3182cf6c6b)

Any insights on how y'all are doing this within the current Module framework would be most appreciated. I have up until this point been using version 3.0.2 which predates the inclusion of Modules. I have been doing POC work on the Netbox open demo site, however.
Thanks.

**Comment by tdentafix** [comment]
Hello,

Apparently the Module model does not allow for nested Modules e.g. module "hierarchy" to be created.
This would however be extremely beneficial to many of us, as you mentioned. We've been using Nokia gear which can have, in the 7750 FP4 lineup, up to 3 levels of cards (that I know of) :

1. The 7750 SR 2/7/14-s always has a first card call "XCM" (eXpandable Control Module). This card has no interfaces but a lot of processing power and is typically tracked in a company's inventory
2. Then into the XCM you can either insert :
* an MDA, directly (Media Dependant Adapter), that has interfaces (e.g. 100GE/400GE/800GE)
* or an IOM, that can in turn house MDAs (typically lower speeds e.g. 25GE)

For the moment we still can't model this properly due to the model limitations, and finding workarounds is risky as it may lead to data migrations once the model has evolved to support this kind of use-case ...

**Comment by mtinberg** [comment]
I would be less worried about needing​ to migrate if the data model extends to support this case, if you want​ to migrate you can plan to do so but shouldn't need​ to, whatever conventions you develop to sufficiently model your use case should continue to work even if there is a better built-in or plugin developed later.

With the tools as they stand today a way to model this is to create precomposed Modules that have the standard build-out of sub-modules how you deploy them, so you end up with a​ module with more or less the right interfaces in it, even if the DB isn't a perfect 1:1 representation of the topology of the hardware.  Maybe that means you have 3 or 4 definitions for XCM modules with different MDA/IOM configurations, but that should be a manageable thing and easy for your network engineers to use, right?


—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Comment by mmfreitas** [comment, marked as answer]
We have a workaround that we use on our ASR9K's. I simply added the sub-slots as module-bays:

![image](https://github.com/netbox-community/netbox/assets/91552302/e2afb845-f081-4914-9f0c-c1f4aca1dbe7)

Then only the submodules have interfaces assigned to its module type allowing them to be installed in the correct position:
![image](https://github.com/netbox-community/netbox/assets/91552302/979c264b-0e8f-43e9-b09c-ef1894f22236)

While the line card does not have any components. This way I can represent the names of components accuratly but unfortunately have to do this by device.

Hope this helps.

**Comment by MattDRogers** [comment]
Thanks everyone for chiming in on this, I really appreciate it. I wanted to make sure I wasn't missing something really obvious in the newer version.

I'm fine with creating Module Types and then populating the interfaces manually or renaming them. So far in v3.0.2 I had just been instantiating my routers as Device Types with only the Console Ports, Management Interfaces (from controller cards) and Power Ports as part of the overall "template." Then, I would manually input the Interfaces based on the cards and sub-cards installed. That method works just fine. I really can't complain. Incorporating Modules once I upgrade would be really cool, though. I can handle a little manual entry and a degree of abstraction in the meantime.

I'm a network guy and not very adept at writing software, or I would try to crack the code on having a responsive middle variable for the automatic/inherited Interface naming.

This community is awesome and Netbox has been a life-saver! Thanks again.

**Comment by mtinberg** [comment]
I'm a network guy and not very adept at writing software, or I would try to crack the code on having a responsive middle variable for the automatic/inherited Interface naming.
https://demo.netbox.dev/static/docs/models/dcim/moduletype/

Is Automatic Component Renaming what you are looking for?

—

Mark Tinberg ***@***.***>

Division of Information Technology-Network Services

University of Wisconsin-Madison


________________________________

**Comment by MattDRogers** [comment]
@mtinberg , yes. And I can see how it works for the example given in the documentation. I have been able to get this to work myself. It is the sub-module's position that's giving me grief.

Example from the documentation: `Gi{module}/0/[1-48]`
{module} will grab the Module Bay's Position and insert it into the naming convention, but there does not appear to be a way (currently) to use a similar variable for the position within the Module for a sub-module (In the example, this would correspond to the "0").

Something like: `Gi{module}/{sub-module}/[1-48]`

So, a little abstraction and manual input appears to be necessary. Which is fine with me.

**Comment by BYT3M3** [comment]
It appears nested modules have been introduced in version 4.1 https://github.com/netbox-community/netbox/pull/16983
